api/handlers.go

```diff
@@ -41,9 +52,47 @@ type (
 
 // healthGET returns the status of the service
 func (api *API) healthGET(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	// The public status that we'll return as response to this call.
+	status := &HealthGET{
+		DBAlive: true,
+	}
+	// Extended health status that we'll log for the benefit of the service's
+	// administrators.
+	extHealth := ExtendedHealth{
+		Health:                   status,
+		NumberSessionsInProgress: api.staticDB.NumberSessionsInProgress(),
+	}
+	// Ensure that we log the extended health information after we gather as
+	// much of it as possible.
+	defer func() {
+		b, err := json.Marshal(extHealth)
+		if err != nil {
+			api.staticLogger.Warnf("Failed to serialize extended health information. Error: %v", err)
+		}
+		api.staticLogger.Info(string(b))
+	}()
+	err := api.staticDB.Ping(req.Context())
+	if err != nil {
+		status.DBAlive = false
+		status.Error = err.Error()
+		api.WriteJSON(w, status)
+		return
+	}
+	hello, err := api.staticDB.Hello(req.Context())
+	if err != nil {
+		status.Error = err.Error()
+		api.WriteJSON(w, status)
+		return
+	}
+	extHealth.Hello = hello
+
 	mp, err := conf.MinPinners(req.Context(), api.staticDB)
-	var status HealthGET
-	status.DBAlive = err == nil
+	if err != nil {
+		status.Error = err.Error()
+		status.Primary = hello.Primary
```

Comment:
this looks like primary will only be available if ping and hello succeeded - let's add it in case we get any error

Reply:
If the `ping` doesn't succeed then we don't have a live connection to the DB and we can't run the `hello`. If we can't run the `hello`, we can't find the primary.

Reply:
that's not entirely true since we've seen a case where ping would fail on one service but reading health check succeeded on the other - those do not seem to be mutually exclusive, probably because ping has 5s timeout and read doesn't so read can succeed even if ping failed right?

Reply:
right, ping does have a short timeout. maybe this is something we should change as well?

Reply:
Anyway, I changed the logic to always try to get all the information it possibly can. This might often result in multiple repeated errors in the error string but it will be more informative than what we had previously.